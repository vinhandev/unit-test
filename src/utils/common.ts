export function checkEven(countParam: number) {
    if (Number.isInteger(countParam) === false) {
      throw new Error("countParam must be a integer");
      
    }
    return countParam % 2 === 0;
  }
  
  export function forEach(items, callback) {
    for (let index = 0; index < items.length; index++) {
      callback(items[index]);
    }
  }