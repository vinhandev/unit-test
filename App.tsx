import { checkEven } from '@/utils/common';
import { StatusBar } from 'expo-status-bar';
import { useState } from 'react';
import { Button, StyleSheet, Text, View } from 'react-native';

export default function App() {
  const [count, setCount] = useState(0);

  return (
    <View style={styles.container}>
      <StatusBar style="auto" />
      <Text>{`Count: ${count}`}</Text>
      <Text testID='status'>{checkEven(count) ? 'Even' : 'Odd'}</Text>
      <Text testID='status'>{checkEven(count) ? 'Even' : 'Odd'}</Text>
      <Button
        testID="button"
        color={'blue'}
        title="Increment"
        onPress={() => setCount(count + 1)}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
