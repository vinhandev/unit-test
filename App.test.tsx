import { checkEven, forEach } from '@/utils/common';
import * as utils from '@/utils/common';
import { fireEvent, render } from '@testing-library/react-native';
import App from './App';

const mockCallback = jest.fn((x) => 42 + x);

describe('<App />', () => {
  // test function
  it('Trigger function isEven success', () => {
    const mockNumber = 20;
    const result = checkEven(mockNumber);
    expect(result).toBe(true);
  });
  it('Trigger function isEven failed', () => {
    const mockNumber = 19;
    const result = checkEven(mockNumber);
    expect(result).toBe(false);
  });
  it('Trigger function no check if is a decimal', () => {
    const mockNumber = 9.5;
    expect(() => {
      checkEven(mockNumber);
    }).toThrow(Error);
  });

  it('forEach mock function', () => {
    forEach([0, 1], mockCallback);

    // The mock function was called twice
    expect(mockCallback.mock.calls).toHaveLength(2);

    // The first argument of the first call to the function was 0
    expect(mockCallback.mock.calls[0][0]).toBe(0);

    // The first argument of the second call to the function was 1
    expect(mockCallback.mock.calls[1][0]).toBe(1);

    // The return value of the first call to the function was 42
    expect(mockCallback.mock.results[0].value).toBe(42);
  });

  it('forEach checkEven function', () => {
    const mockCheckEven = jest.fn(checkEven);
    forEach([0, 1], mockCheckEven);
    // The mock function was called twice
    expect(mockCheckEven.mock.calls).toHaveLength(2);

    // The first argument of the first call to the function was 0
    expect(mockCheckEven.mock.calls[0][0]).toBe(0);

    // The first argument of the second call to the function was 1
    expect(mockCheckEven.mock.calls[1][0]).toBe(1);

    // The return value of the first call to the function was 42
    expect(mockCheckEven.mock.results[0].value).toBe(true);
  });
  it('spyOn function function', () => {
    const { getByTestId } = render(<App />);
    const spy = jest.spyOn(utils, 'checkEven');
    const button = getByTestId('button');
    fireEvent(button, 'press');
    expect(spy).toHaveBeenCalledWith(1);
    expect(utils.checkEven(1)).toBe(false);
  });
  it('testing component ', () => {
    const { getByTestId } = render(<App />);
    const button = getByTestId('button');
    expect(button).toBeDefined();
  });
  it('match snapshot ', () => {
    const app = render(<App />);

    expect(app).toMatchSnapshot();
  });
  it('layout ', () => {
    const app = render(<App />);
    const button = app.getByTestId('button');

    expect(button.props.children[0].props.children.props.children).toBe('Increment');
  });
});
