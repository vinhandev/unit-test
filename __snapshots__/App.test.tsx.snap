// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> match snapshot  1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Count: 0
  </Text>
  <Text
    testID="status"
  >
    Even
  </Text>
  <Text
    testID="status"
  >
    Even
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="button"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
            {
              "color": "blue",
            },
          ]
        }
      >
        Increment
      </Text>
    </View>
  </View>
</View>
`;
